### ISSUE TITLE
"Model name conflict" when embedded objects share same base name

### ISSUE THREAD

**Issue by iann0036** [aws-cloudformation/cloudformation-cli#406]
When subobjects within an object share the same parent name, their values conflict when the rewriting occurs. This should ideally generate a definition name that doesn't conflict with the other (perhaps with an iterator?).

Example debug log and schema attached.
[debug-log.txt](https://github.com/aws-cloudformation/cloudformation-cli/files/4348176/debug-log.txt)
[terraform-aws-alblistener.json.txt](https://github.com/aws-cloudformation/cloudformation-cli/files/4348179/terraform-aws-alblistener.json.txt)

**Comment by iann0036** [comment]
I'm currently using this in resolver.py:

```
    def _get_model_name_from_ref(self, ref_path):
        """Given a json schema ref, returns the best guess at a model name."""
        if ref_path == ():
            return self._base_model_name

        class_name_iterator = 1  # deliberate; 2 means second model of same name

        class_name = base_class_from_ref(ref_path)

        while class_name_iterator < 99:
            try:
                dupe_path = next(
                    path for path, name in self._models.items() if name == class_name
                )
            except StopIteration:
                return class_name
            class_name_iterator += 1
            class_name = base_class_from_ref(ref_path) + str(class_name_iterator)

        raise ModelResolverError(
            "Model name conflict. "
            f"'{class_name}' found at {dupe_path} and {ref_path}"
        )
```

which seems to be working well. I have no idea how the resolver tests work so would rely on someone else for coverage.

**Comment by ammokhov** [comment]
```json
        "AuthenticateCognito": {
            "type": "array",
            "insertionOrder": false,
            "items": {
                "$ref": "#/definitions/AuthenticateCognito"
            },
            "maxItems": 1
        }
```

has circular ref, this should be fixed by the following PR #596

**Comment by iann0036** [comment]
@ammokhov,

FYI, that looked like an unrelated issue (I'm only just getting back to this now). See updated example model:

```
{
    "typeName": "My::Test::Resource",
    "description": "Example",
    "sourceUrl": "https://example.com/",
    "documentationUrl": "https://example.com/",
    "definitions": {},
    "properties": {
        "GenericID": {
            "description": "Generic ID",
            "type": "string"
        },
        "PropertyA": {
            "type": "array",
            "insertionOrder": false,
            "items": {
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "Items": {
                        "type": "array",
                        "insertionOrder": false,
                        "items": {
                            "type": "object",
                            "additionalProperties": false,
                            "properties": {
                                "Subprop": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "PropertyB": {
            "type": "array",
            "insertionOrder": false,
            "items": {
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "Items": {
                        "type": "array",
                        "insertionOrder": false,
                        "items": {
                            "type": "object",
                            "additionalProperties": false,
                            "properties": {
                                "Subprop": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "additionalProperties": false,
    "readOnlyProperties": [
        "/properties/GenericID"
    ],
    "primaryIdentifier": [
        "/properties/GenericID"
    ],
    "handlers": {
        "create": {
            "permissions": [
                "*"
            ]
        },
        "read": {
            "permissions": [
                "*"
            ]
        },
        "update": {
            "permissions": [
                "*"
            ]
        },
        "delete": {
            "permissions": [
                "*"
            ]
        },
        "list": {
            "permissions": [
                "*"
            ]
        }
    }
}
```
